File: src/main/java/ru/ioleksiv/telegram/bot/core/controller/annotations/parser/finder/ValidatorFinder.java
package ru.ioleksiv.telegram.bot.core.controller.annotations.parser.finder;

import org.springframework.stereotype.Controller;
import ru.ioleksiv.telegram.bot.core.controller.annotations.holder.CustomValidatorHolder;
import ru.ioleksiv.telegram.bot.core.controller.annotations.parser.filter.FilterParser;
import ru.ioleksiv.telegram.bot.core.controller.handler.check.Validator;
import ru.ioleksiv.telegram.bot.core.model.ITelegram;

import java.lang.annotation.Annotation;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

@Controller
public class ValidatorFinder implements Finder {

    private final Map<Class, FilterParser> annotationParserMap = new HashMap<>();
    private final CustomValidatorHolder customValidatorHolder;

    public ValidatorFinder(Collection<FilterParser> filterParsers,
                           CustomValidatorHolder customValidatorHolder) {

        for (FilterParser filterParser : filterParsers) {
            annotationParserMap.put(filterParser.getAnnotationClass(), filterParser);
        }

        this.customValidatorHolder = customValidatorHolder;

    }

    @Override
    public <SA extends Annotation, T> Validator<T> find(SA annotation) {
        Class<? extends Annotation> keyClass = annotation.annotationType();

        if (!annotationParserMap.containsKey(keyClass)) {
            return argument -> true;
        }

        FilterParser<SA, T> secondaryFilterParser = (FilterParser<SA, T>) annotationParserMap.get(keyClass);
        return secondaryFilterParser.createChecker(annotation, this);
    }

    @Override
    public <T extends ITelegram> Validator<T> find(String validatorName, Class<T> targetClass) {
        return customValidatorHolder.get(validatorName, targetClass);
    }
}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/finder/ValidatorFinder.kt
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.finder

import io.github.oleksivio.telegram.bot.core.controller.annotations.holder.CustomValidatorHolder
import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.filter.FilterParser
import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator
import io.github.oleksivio.tl.kbot.server.api.model.ITelegram
import org.springframework.stereotype.Controller
import kotlin.reflect.KClass

/**
 * Created by oleksivio on 15.03.19 at 9:44
 * Project: telegram-bot-api
 */
@Controller
class ValidatorFinder(filterParsers: Collection<FilterParser<*, *>>,
                      private val customValidatorHolder: CustomValidatorHolder) : Finder {

    private val annotationParserMap = hashMapOf<KClass<*>, FilterParser<*, *>>()

    init {
        for (filterParser in filterParsers) {
            annotationParserMap[filterParser.parserAnnotationClass] = filterParser
        }
    }

    override fun <SA : Annotation, T> find(annotation: SA): Validator<T> {
        val keyClass = annotation.annotationClass

        if (!annotationParserMap.containsKey(keyClass)) {
            return { true }
        }

        val secondaryFilterParser = annotationParserMap[keyClass] as FilterParser<SA, T>

        return secondaryFilterParser.createChecker(annotation, this)
    }

    override fun <T : ITelegram> find(validatorName: String, targetClass: KClass<T>): Validator<T> {
        return customValidatorHolder[validatorName, targetClass]
    }
}
